core/templates/domain/exploration/RuleInputTypeFactory.ts

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory to determine correct type of RuleInputs
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { IRuleInput } from 'domain/exploration/RuleObjectFactory';
+import { INumberWithUnitsBackendDict } from
+  'domain/objects/NumberWithUnitsObjectFactory';
+import { IFractionDict } from
+  'domain/objects/FractionObjectFactory.ts';
+/* eslint-disable max-len */
+import { INote } from 'extensions/interactions/MusicNotesInput/directives/music-notes-input-rules.service';
+/* eslint-enable max-len */
+import { IGraphBackendDict } from
+  'extensions/interactions/GraphInput/directives/graph-detail.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RuleInputTypeFactory {
+  _isGraph(variable: Object): variable is IGraphBackendDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'vertices' in variable;
+  }
+
+  _isNoteArray(variable: Object): variable is INote[] {
+    if (Array.isArray(variable)) {
+      if (variable.length === 0) {
+        return true;
+      }
+
+      if (typeof variable[0] === 'string' || typeof variable[0] === 'number') {
+        return false;
+      }
+
+      return 'readableNoteName' in variable[0];
```

Comment:
Similar  comment, the field may be there in variable[0] but may not be in other elements, so make sure to check all elements.
Also, why aren't you verifying whether all the fields are there, why just this one?

Reply:
Same reply as above.